client/asset/dcr/dcr.go

```diff
@@ -5762,93 +5832,404 @@ func (dcr *ExchangeWallet) txDB() *btc.BadgerTxDB {
+				continue
 			}
 
-			dcr.addTxToHistory(txType, txHash, toAtoms(tx.Amount), fee, nil, nil, true)
+			addr := addrs[0]
+			owns, err := dcr.walletOwnsAddress(addr)
+			if err != nil {
+				dcr.log.Errorf("walletOwnsAddress error: %w", err)
+				continue
+			}
+			if owns {
+				in += uint64(txOut.Value)
+			} else {
+				out += uint64(txOut.Value)
+			}
 		}
+		return
+	}
 
-		for _, tx := range recentTxs {
-			if *tx.TxType == walletjson.LTTTRegular && tx.Category == "receive" {
-				addToHistoryIfNew(&tx, asset.Receive)
-			} else if *tx.TxType == walletjson.LTTTVote {
-				addToHistoryIfNew(&tx, asset.TicketVote)
-			} else if *tx.TxType == walletjson.LTTTRevocation {
-				addToHistoryIfNew(&tx, asset.TicketRevocation)
+	in, out := txOutDirection(msgTx)
+
+	getRecipient := func(msgTx *wire.MsgTx, receive bool) *string {
+		for _, txOut := range msgTx.TxOut {
+			_, addrs := stdscript.ExtractAddrs(scriptVersion, txOut.PkScript, dcr.chainParams)
+			if err != nil {
+				dcr.log.Errorf("ExtractAddrs error: %w", err)
+				continue
+			}
+			if len(addrs) != 1 { // sanity check
+				continue
+			}
+
+			addr := addrs[0]
+			owns, err := dcr.walletOwnsAddress(addr)
+			if err != nil {
+				dcr.log.Errorf("walletOwnsAddress error: %w", err)
+				continue
+			}
+
+			if receive == owns {
+				str := addr.String()
+				return &str
 			}
 		}
+		return nil
+	}
+
+	if tx.Send {
+		txType := asset.Send
+		if allOutputsPayUs(msgTx) {
+			txType = asset.SelfSend
+		}
+		return &asset.WalletTransaction{
+			Type:      txType,
+			ID:        tx.TxID,
+			Amount:    out,
+			Fees:      fee,
+			Recipient: getRecipient(msgTx, false),
+		}, nil
+	}
+
+	return &asset.WalletTransaction{
+		Type:      asset.Receive,
+		ID:        tx.TxID,
+		Amount:    in,
+		Fees:      fee,
+		Recipient: getRecipient(msgTx, true),
+	}, nil
+}
+
+// addUnknownTransactionsToHistory checks for any transactions the wallet has
+// made or recieved that are not part of the transaction history. It scans
+// from the last point to which it had previously scanned to the current tip.
+func (dcr *ExchangeWallet) addUnknownTransactionsToHistory(tip uint64) {
+	txHistoryDB := dcr.txDB()
+
+	const blockQueryBuffer = 3
+	var blockToQuery uint64
+	lastQuery := dcr.receiveTxLastQuery.Load()
+	if lastQuery == 0 {
+		// TODO: use wallet birthday instead of block 0.
+		// blockToQuery = 0
+	} else if lastQuery < tip-blockQueryBuffer {
+		blockToQuery = lastQuery - blockQueryBuffer
+	} else {
+		blockToQuery = tip - blockQueryBuffer
+	}
+
+	txs, err := dcr.wallet.ListSinceBlock(dcr.ctx, int32(blockToQuery))
+	if err != nil {
+		dcr.log.Errorf("Error listing transactions since block %d: %v", blockToQuery, err)
+		return
+	}
+
+	for _, tx := range txs {
+		txHash, err := chainhash.NewHashFromStr(tx.TxID)
+		if err != nil {
+			dcr.log.Errorf("Error decoding tx hash %s: %v", tx.TxID, err)
+			continue
+		}
+		_, err = txHistoryDB.GetTx(txHash.String())
+		if err == nil {
+			continue
+		}
```

Comment:
I was adding this to zec and noticed here that the ListTransactions result looks like this:
```
Result:
{
 "transactions": [{                 (array of object) JSON array of objects containing verbose details of the each transaction
  "account": "value",               (string)          DEPRECATED -- Unset
  "address": "value",               (string)          Payment address for a transaction output
  "amount": n.nnn,                  (numeric)         The value of the transaction output valued in decred
  "blockhash": "value",             (string)          The hash of the block this transaction is mined in, or the empty string if unmined
  "blockindex": n,                  (numeric)         Unset
  "blocktime": n,                   (numeric)         The Unix time of the block header this transaction is mined in, or 0 if unmined
  "category": "value",              (string)          The kind of transaction: "send" for sent transactions, "immature" for immature coinbase outputs, "generate" for mature coinbase outputs, or "recv" for all other received outputs.  Note: A single output may be included multiple times under different categories
  "confirmations": n,               (numeric)         The number of block confirmations of the transaction
  "fee": n.nnn,                     (numeric)         The total input value minus the total output value for sent transactions
  "generated": true|false,          (boolean)         Whether the transaction output is a coinbase output
  "involveswatchonly": true|false,  (boolean)         Unset
  "time": n,                        (numeric)         The earliest Unix time this transaction was known to exist
  "timereceived": n,                (numeric)         The earliest Unix time this transaction was known to exist
  "txid": "value",                  (string)          The hash of the transaction
  "txtype": "value",                (string)          The type of tx (regular tx, stake tx)
  "vout": n,                        (numeric)         The transaction output index
  "walletconflicts": ["value",...], (array of string) Unset
  "comment": "value",               (string)          Unset
  "otheraccount": "value",          (string)          Unset
 },...],                                              
 "lastblock": "value",              (string)          Hash of the latest-synced block to be used in later calls to listsinceblock
}    
```

This is one vout of a transaction, what if a tx had two outputs paying to the wallet? Would that txid then appear twice? In that case the second txid would be passed over.

Reply:
I guess it's fine with `idUnknownTx` which takes the txid as a whole. Nevermind.